### PR TITLE
Allow boxes to move like other pointer types, add tests

### PR DESCRIPTION
**Pull request by aidanhs**
Not sure how comprehensive you prefer tests to be - I can remove the value checks if you don't think they add anything.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -144,7 +144,7 @@ fn is_ty_no_move(tcx: &ty::ctxt, t: ty::Ty) -> bool {
                 }
                 true
             },
-            ty::TyRef(..) | ty::TyRawPtr(..) => false, // don't recurse down ptrs
+            ty::TyBox(..) | ty::TyRef(..) | ty::TyRawPtr(..) => false, // don't recurse down ptrs
```
Comment (Manishearth) [inline src/lib.rs]: Box is an owning ref, the other two are not. The reason we don't recurse here is that the other two are not owned references so cannot change the underlying ownership of the thing being pointed to.

On the other hand, box is owned, and moving it is a change of ownership.


This crate considers a move to be any change in ownership, not just an in-memory move.

**Comment by Manishearth** [comment]
This doesn't align with the goals of this crate as listed above. You can of course create a `MovableBox` struct that does allow this to work around it.

**Comment by aidanhs** [comment]
For my clarity - this is because once you've created a box you can move the value back out of it, effectively subverting this lint?

**Comment by Manishearth** [comment]
That's a part of it, but not the main reason -- the main reason is that this crate isn't just about in-memory moves, it's about conceptual moves too (Servo needs this)

**Comment by aidanhs** [comment]
Ah I see, thanks.

**Pull request closed without merging**